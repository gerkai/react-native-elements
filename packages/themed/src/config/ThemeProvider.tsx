import React, { useEffect } from 'react';
import deepmerge from 'deepmerge';
import { colors, darkColors, Colors } from './colors';
import { Theme, ThemeMode, RecursivePartial } from './theme';

export type { RecursivePartial };

export type CreateTheme = Theme<{
  darkColors?: RecursivePartial<Colors>;
}>;

export type UpdateTheme = (
  myNewTheme: CreateTheme | ((myTheme: Theme) => CreateTheme)
) => void;

export type ReplaceTheme = (
  updates: CreateTheme | ((myTheme: Theme) => CreateTheme)
) => void;

export type ThemeProps<T = {}> = {
  theme: Theme<T>;
  updateTheme: UpdateTheme;
  replaceTheme: ReplaceTheme;
};

type ProviderProps = {
  theme?: CreateTheme;
};

export const ThemeContext: React.Context<ThemeProps> = React.createContext({
  theme: {
    colors,
  },
} as ThemeProps);

export const createTheme = (theme: CreateTheme): CreateTheme => {
  return deepmerge({ colors, darkColors, mode: 'light' } as CreateTheme, theme);
};

const separateColors = (theme: CreateTheme, themeMode?: ThemeMode): Theme => {
  const {
    darkColors: themeDarkColors = {},
    colors: themeLightColors = {},
    mode = themeMode,
    ...restTheme
  } = theme;

  const themeColors = mode === 'dark' ? themeDarkColors : themeLightColors;
  return { colors: themeColors, mode, ...restTheme };
};

/**
 * const myTheme = createTheme({
 *  mode: 'light'|'dark',
 *  colors: {},
 *  darkColors: {},
 * });
 *
 * <ThemeProvider theme={myTheme}>
 *   <MyComponent />
 * </ThemeProvider>
 *
 */
export const ThemeProvider: React.FC<ProviderProps> = ({
  theme = {},
  children,
}) => {
  const [themeState, setThemeState] = React.useState<CreateTheme>(theme);
  const isMounted = React.useRef<boolean>(false);

  useEffect(() => {
    if (isMounted.current) {
      setThemeState(theme);
    } else {
      isMounted.current = true;
    }
  }, [theme.mode, theme.colors, theme.darkColors, theme]);

  const updateTheme: UpdateTheme = React.useCallback((updatedTheme) => {
    setThemeState((oldTheme) => {
      const newTheme =
        typeof updatedTheme === 'function'
          ? updatedTheme(oldTheme)
          : updatedTheme;
      return deepmerge({ ...oldTheme }, newTheme);
    });
  }, []);

  const replaceTheme: ReplaceTheme = React.useCallback((replacedTheme) => {
    setThemeState((oldTheme) => {
      const newTheme =
        typeof replacedTheme === 'function'
          ? replacedTheme(oldTheme)
          : replacedTheme;
      return deepmerge(createTheme({}), newTheme);
    });
  }, []);

  return (
    <ThemeContext.Provider
      value={{
        theme: separateColors(themeState),
        updateTheme,
        replaceTheme,
      }}
    >
      {children}
    </ThemeContext.Provider>
  );
};

export const ThemeConsumer = ThemeContext.Consumer;
