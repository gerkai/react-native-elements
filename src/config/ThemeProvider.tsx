import React from 'react';
import deepmerge from 'deepmerge';
import colors from './colors';
import darkColors from './colorsDark';
import { FullTheme, Theme } from './theme';

type RecursivePartial<T> = { [P in keyof T]?: RecursivePartial<T[P]> };

export interface ThemeProps<T> {
  theme: Theme<T>;
  updateTheme: (updates: RecursivePartial<FullTheme>) => void;
  replaceTheme: (updates: RecursivePartial<FullTheme>) => void;
}

export const ThemeContext: React.Context<ThemeProps<{}>> = React.createContext({
  theme: {
    colors,
  },
} as ThemeProps<{}>);

export type ThemeProviderProps = {
  useDark?: boolean;
};

type ThemeProviderState = {
  theme: Theme;
  useDark: boolean;
};

export default class ThemeProvider extends React.Component<
  ThemeProviderProps,
  ThemeProviderState
> {
  static defaultProps = {
    theme: {},
    useDark: false,
  };
  defaultTheme: Partial<FullTheme>;

  constructor(props: { theme: Theme; useDark?: boolean }) {
    super(props);
    const defaultColors = props.useDark ? darkColors : colors;
    this.defaultTheme = deepmerge(
      {
        colors: defaultColors,
      },
      props.theme
    );
    this.state = {
      theme: this.defaultTheme,
      useDark: Boolean(props.useDark),
    };
  }

  static getDerivedStateFromProps(
    props: {
      theme: Theme;
      useDark?: boolean;
    },
    state: ThemeProviderState
  ) {
    const { useDark } = props;
    if (useDark !== state.useDark) {
      const defaultColors = useDark ? darkColors : colors;
      return {
        theme: deepmerge(
          state.theme,
          deepmerge(
            {
              colors: defaultColors,
            },
            props.theme
          )
        ),
        useDark,
      };
    }
    return null;
  }

  updateTheme = (updates: RecursivePartial<FullTheme>) => {
    this.setState(({ theme }) => ({
      theme: deepmerge(theme, updates),
    }));
  };

  replaceTheme = (theme: RecursivePartial<FullTheme>) => {
    this.setState(() => ({
      theme: deepmerge(this.defaultTheme, theme),
    }));
  };

  getTheme = () => this.state.theme;

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.state.theme,
          updateTheme: this.updateTheme,
          replaceTheme: this.replaceTheme,
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export const ThemeConsumer = ThemeContext.Consumer;
