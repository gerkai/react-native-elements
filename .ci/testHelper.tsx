import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider, FullTheme } from '../src/config';

const theme = {
  colors: {
    black: 'black',
  },
};

export const renderWithTheme = (
  children: any,
  themeProp: Partial<FullTheme> = theme
) => {
  return render(<ThemeProvider theme={themeProp}>{children}</ThemeProvider>);
};
