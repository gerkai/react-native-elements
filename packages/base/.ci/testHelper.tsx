import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import { ThemeProvider, FullTheme, colors } from '../src/config';
import deepmerge from 'deepmerge';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<FullTheme> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    themeProp ? (
      <ThemeProvider theme={deepmerge({ colors }, themeProp)}>
        {children}
      </ThemeProvider>
    ) : (
      children
    ),
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper');
  return { wrapper, ...renderApi };
};
