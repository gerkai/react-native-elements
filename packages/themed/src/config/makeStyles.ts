import { useContext, useMemo } from 'react';
import { StyleSheet } from 'react-native';
import { Colors } from './colors';
import { FullTheme, Theme, ThemeMode } from './theme';
import { ReplaceTheme, ThemeContext, UpdateTheme } from './ThemeProvider';

interface useThemeI {
  replaceTheme: ReplaceTheme;
  updateTheme: UpdateTheme;
  theme: {
    colors: Colors;
    mode: ThemeMode;
  } & Theme;
}

export const useTheme = (): useThemeI => {
  return useContext(ThemeContext);
};

export const useThemeMode = () => {
  return useContext(ThemeContext).theme.mode;
};

export const makeStyles =
  <T extends StyleSheet.NamedStyles<T> | StyleSheet.NamedStyles<any>, V>(
    styles: T | ((theme: Partial<FullTheme>, props: V) => T)
  ) =>
  (props: V = {} as any): T => {
    const { theme } = useTheme();

    return useMemo(() => {
      const css =
        typeof styles === 'function'
          ? styles({ colors: theme.colors, mode: theme.mode }, props)
          : styles;
      return StyleSheet.create(css);
    }, [props, theme.colors, theme.mode]);
  };
