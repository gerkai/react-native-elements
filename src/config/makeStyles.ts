import { useContext } from 'react';
import { StyleSheet } from 'react-native';
import { FullTheme } from './theme';
import { ThemeContext } from './ThemeProvider';

export const useTheme = () => {
  return useContext(ThemeContext);
};

export const makeStyles = <
  T extends StyleSheet.NamedStyles<T> | StyleSheet.NamedStyles<any>,
  V
>(
  styles: T | ((theme: Partial<FullTheme>, props: V) => T)
) => (props: V = {} as any): T => {
  const { theme } = useTheme();
  const css = typeof styles === 'function' ? styles(theme, props) : styles;

  return StyleSheet.create(css);
};
