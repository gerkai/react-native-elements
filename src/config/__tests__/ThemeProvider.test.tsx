import React from 'react';
import { Text, Button } from '../..';
import DefaultTheme, { FullTheme } from '../theme';
import { renderWithWrapper } from '../../../.ci/testHelper';
import { fireEvent, render } from '@testing-library/react-native';
import { useTheme } from '../makeStyles';
import ThemeProvider from '../ThemeProvider';
import { View } from 'react-native';

describe('ThemeProvider', () => {
  it('should work', () => {
    const { toJSON } = renderWithWrapper(<Text />);
    expect(toJSON).toMatchSnapshot();
  });

  it('should update and replace theme', () => {
    const TestComp = (): JSX.Element => {
      const { theme, replaceTheme, updateTheme } = useTheme();
      return (
        <>
          <Button
            testID="updateTheme"
            onPress={() => {
              updateTheme({
                colors: {
                  primary: 'red',
                },
              });
            }}
          />
          <Button
            testID="replaceThemeButton"
            onPress={() => {
              replaceTheme({
                colors: {
                  primary: 'purple',
                },
              });
            }}
          />
          <Text testID="themeChild" children={theme.colors.primary} />
        </>
      );
    };

    const { queryByTestId } = render(
      <ThemeProvider>
        <TestComp />
      </ThemeProvider>
    );
    const updateButton = queryByTestId('updateTheme');
    const replaceButton = queryByTestId('replaceThemeButton');
    const textTheme = queryByTestId('themeChild');
    expect(textTheme.props.children).toEqual(DefaultTheme.colors.primary);

    fireEvent.press(updateButton);
    expect(textTheme.props.children).toEqual('red');

    fireEvent.press(replaceButton);
    expect(textTheme.props.children).toBe('purple');
  });

  it('should contain useDark key', () => {
    const { queryByTestId } = renderWithWrapper(
      <ThemeProvider theme={DefaultTheme} useDark>
        <View testID="viewComp" />
      </ThemeProvider>
    );
    const instance = queryByTestId('viewComp');
    expect(instance.parent.parent.props).toMatchObject({
      theme: DefaultTheme,
      useDark: true,
    });
  });

  it('should contain useDark key with value true', () => {
    const { toJSON } = renderWithWrapper(
      <ThemeProvider useDark={true}>
        <View />
      </ThemeProvider>
    );
    expect(toJSON).toMatchSnapshot();
  });

  it('should retain custom theme when switching between light / dark mode', () => {
    const customTheme: Partial<FullTheme> = {
      Text: { accessibilityLabel: 'theme-test' },
    };
    const TestComp = (): JSX.Element => {
      const [useDark, setUseDark] = React.useState(false);
      return (
        <>
          <Button testID="updateTheme" onPress={() => setUseDark(!useDark)} />
          <ThemeProvider theme={customTheme} useDark={useDark}>
            <Text testID="themeChild" children={String(useDark)} />
          </ThemeProvider>
        </>
      );
    };
    const { queryByTestId } = render(<TestComp />);
    const updateButton = queryByTestId('updateTheme');
    const textTheme = queryByTestId('themeChild');
    const themeProvider = queryByTestId('themeChild').parent.parent.parent
      .parent;

    expect(textTheme.props.accessibilityLabel).toEqual('theme-test');

    // set to dark mode
    fireEvent.press(updateButton);
    expect(themeProvider.props.useDark).toBeTruthy();
    expect(textTheme.props.accessibilityLabel).toEqual('theme-test');

    // set to light mode
    fireEvent.press(updateButton);
    expect(themeProvider.props.useDark).toBeFalsy();
    expect(textTheme.props.accessibilityLabel).toEqual('theme-test');
  });
});
