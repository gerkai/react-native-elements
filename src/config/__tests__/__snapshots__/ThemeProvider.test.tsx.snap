// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider should contain useDark key with value true 1`] = `[Function]`;

exports[`ThemeProvider should work 1`] = `[Function]`;
